x/ccv/provider/keeper/relay.go

```diff
@@ -211,9 +232,14 @@ func (k Keeper) EndBlockCIS(ctx sdk.Context) {
 	valUpdateID := k.GetValidatorSetUpdateId(ctx)
 	// set the ValsetUpdateBlockHeight
 	k.SetValsetUpdateBlockHeight(ctx, valUpdateID, uint64(ctx.BlockHeight()+1))
+	// Execute slash packet throttling logic
+	k.HandlePendingSlashPackets(ctx)
+	// Replenish slash meter if necessary
+	k.CheckForSlashMeterReplenishment(ctx)
```

Comment:
```suggestion
        // Replenish slash meter if necessary
	k.CheckForSlashMeterReplenishment(ctx)
	k.HandlePendingSlashPackets(ctx)
```

Reply:
Please note that, even with this change in ordering, some funky stuff is still possible simply due to the replenish period. This gets sketchier if you consider the CancelUndelegation tx which is coming one day.

The meter might be set a height h to a value x, then the total power shrinks changes due to undelegations, then the valset at h+k will be slashable with a really high meter value.

Reply:
This suggested change looks correct, thanks! I think the newer version of `CheckForSlashMeterReplenishment` accounts for the mentioned scenario by decrementing the meter if the allowance becomes smaller. I'll confirm when things are less crazy

Reply:
Following up on this comment, there's multiple concerns here:
1. [this bug](https://github.com/cosmos/interchain-security/pull/462#discussion_r1044537365) is fixed by changing the ordering of `CheckForSlashMeterReplenishment` and `HandlePendingSlashPackets` as you've suggested. This off-by-one bug could surely cause some odd behavior and potential attack vectors if left unfixed.
2. The practical concern of total voting power changing between blocks. The solution to this concern is the way [CheckForSlashMeterReplenishment](https://github.com/cosmos/interchain-security/blob/5ce8ea28a667f9caf52aab1f986d50b7918d0655/x/ccv/provider/keeper/throttle.go#L120) is designed, as validated by [TestTotalVotingPowerChanges](https://github.com/cosmos/interchain-security/blob/5ce8ea28a667f9caf52aab1f986d50b7918d0655/x/ccv/provider/keeper/throttle_test.go#L291). We need the slash meter to be capped by the current slash allowance. We also need to allow the meter to eventually hit the allowance/max value (after a replenishment) in the event that total voting power increases. 
3. the invariant concerns of total voting power increasing between blocks. This hasn't been adequately addressed beyond the cop-out assumption mentioned above

Reply:
1. is addressed by https://github.com/cosmos/interchain-security/commit/408ff0c18c607259754e1799c45f8c0f2d44476c